src/lib/features/feature-lifecycle/feature-lifecycle-service.ts

```diff
@@ -224,41 +225,52 @@ export class FeatureLifecycleService extends EventEmitter {
     public calculateStageDurations(
         featureLifeCycles: FeatureLifecycleProjectItem[],
     ) {
-        const groupedByFeature = featureLifeCycles.reduce<{
-            [feature: string]: FeatureLifecycleProjectItem[];
-        }>((acc, curr) => {
-            if (!acc[curr.feature]) {
-                acc[curr.feature] = [];
-            }
-            acc[curr.feature].push(curr);
-            return acc;
-        }, {});
+        const sortedLifeCycles = featureLifeCycles.sort(
+            (a, b) => a.enteredStageAt.getTime() - b.enteredStageAt.getTime(),
+        );
 
-        const times: IFeatureLifecycleStageDuration[] = [];
-        Object.values(groupedByFeature).forEach((stages) => {
-            stages.sort(
-                (a, b) =>
-                    a.enteredStageAt.getTime() - b.enteredStageAt.getTime(),
-            );
+        const groupedByProjectAndStage = sortedLifeCycles.reduce<{
+            [key: string]: number[];
+        }>((acc, curr, index, array) => {
+            const key = `${curr.project}/${curr.stage}`;
+            if (!acc[key]) {
+                acc[key] = [];
+            }
 
-            stages.forEach((stage, index) => {
-                const nextStage = stages[index + 1];
-                const endTime = nextStage
-                    ? nextStage.enteredStageAt
-                    : new Date();
-                const duration = differenceInMinutes(
-                    endTime,
-                    stage.enteredStageAt,
+            const nextItem = array
+                .slice(index + 1)
+                .find(
+                    (item) =>
+                        item.feature === curr.feature &&
+                        item.stage !== curr.stage,
                 );
+            const endTime = nextItem ? nextItem.enteredStageAt : new Date();
+            const duration = differenceInMinutes(endTime, curr.enteredStageAt);
+            acc[key].push(duration);
+            return acc;
+        }, {});
 
-                times.push({
-                    feature: stage.feature,
-                    stage: stage.stage,
-                    project: stage.project,
-                    duration,
-                });
+        const medians: IProjectLifecycleStageDuration[] = [];
+        Object.entries(groupedByProjectAndStage).forEach(([key, durations]) => {
+            const [project, stage] = key.split('/');
+            const duration = this.calculateMedian(durations);
+            medians.push({
+                project,
+                stage: stage as StageName,
+                duration,
             });
         });
-        return times;
+
+        return medians;
+    }
+
+    private calculateMedian(numbers: number[]): number {
```

Comment:
I'd move it to median.ts or math.ts in utils and test it separately as a pure function not attached to a class as a private method

Reply:
Done